engine/builtin.go

```diff
@@ -813,6 +813,50 @@ func Compare(order, term1, term2 Term, k func(*Env) *Promise, env *Env) *Promise
 	}
 }
 
+// Between succeeds when lower, upper, and value are all integers, and lower <= value <= upper.
+// If value is a variable, it is unified with successive integers from lower to upper.
+func Between(lower, upper, value Term, k func(*Env) *Promise, env *Env) *Promise {
+	var low, high Integer
+
+	switch lower := env.Resolve(lower).(type) {
+	case Integer:
+		low = lower
+	case Variable:
+		return Error(ErrInstantiation)
+	default:
+		return Error(TypeErrorInteger(lower))
+	}
+
+	switch upper := env.Resolve(upper).(type) {
+	case Integer:
+		high = upper
+	case Variable:
+		return Error(ErrInstantiation)
+	default:
+		return Error(TypeErrorInteger(upper))
+	}
+
+	if low > high {
+		return Bool(false)
+	}
+
+	switch value := env.Resolve(value).(type) {
+	case Integer:
+		if value >= low && value <= high {
+			return k(env)
+		}
+		return Bool(false)
+	case Variable:
+		return Delay(func(context.Context) *Promise {
+			return Unify(value, low, k, env)
+		}, func(context.Context) *Promise {
+			return Between(low+1, upper, value, k, env)
```

Comment:
When `low = high = math.MaxInt64`, `low+1` overflows and becomes a negative integer.

This could be solved by checking if `low < high`:

```go
if low == high { // Since we've already checked low <= high.
	return Bool(false)
}
return Between(low+1, upper, value, k, env)
```

Or, for a better performance:

```go
ks := make([]func(*Env) *Promise, 0, 2)
ks = append(ks, func(context.Context) *Promise {
	return Unify(value, low, k, env)
})
if low < high {
	ks = append(ks, func(context.Context) *Promise {
		return Between(low+1, upper, value, k, env)
	})
}
return Delay(ks...)
```

Reply:
Thanks! I changed it to the better performance suggestion.
Before it would overflow but get caught by 
```go
if low > high {
	return Bool(false)
}
```
but now it should completely eliminate that dead end 👍